### PR TITLE
docs(fern): add multi-turn agent (tool-calling) SFT tutorial

### PR DESCRIPTION
**Pull request by khazic**
# What does this PR do ?

Adds an end-to-end tutorial for **multi-turn agent (tool-calling) SFT** with Qwen2.5-3B on the `llamafactory/glaive_toolcall_en` dataset, using `make_agent_chat_dataset` and the generation-based `ToolCallAccuracyEvaluator`. It follows the before/after, step-by-step structure of the Gemma 4 guide and complements the existing single-turn [FunctionGemma](/recipes-e2e-examples/function-calling) tutorial (which covers one-shot `make_xlam_dataset` SFT).

# Changelog

- New page `fern/versions/nightly/pages/guides/llm/agent-sft.mdx` (Steps 0–6: environment, dataset exploration, base-model eval, training config, launch, fine-tuned eval, results comparison; plus a LoRA/PEFT variant).
- Nav entry added to `fern/versions/nightly.yml` under **Recipes & E2E Examples**, after the FunctionGemma guide (slug: `agent-sft`).
- Teaching callouts for `answer_only_loss_mask`, the `train_on_last_turn_only` / `mask_reasoning_content` / `drop_history_reasoning_content` flags, the `seq_length` + `truncation` requirement, and the FSDP2 in-loop-eval caveat.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests? — N/A (docs-only)
- [x] Did you add or update any necessary documentation?

# Additional Information

- Added to the **nightly** docs tree only (where new pages land); the frozen v0.4 GA aliases are untouched (a back-port is a separate, deliberate step).
- **Training log and result metrics are illustrative placeholders** (marked as such in the page) to show the expected base-low / fine-tuned-high shape — they should be replaced with real run output before this is treated as published guidance.
- `fern check` runs in the docs CI; a 🌿 Fern preview URL is posted automatically for `fern/**` changes.

### CODE REVIEW

File: fern/versions/nightly/pages/guides/llm/agent-sft.mdx [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+| **Step 1** | Explore the `glaive_toolcall` dataset |
+| **Step 2** | Evaluate the base model (before fine-tuning) |
+| **Step 3** | Training configuration |
+| **Step 4** | Launch fine-tuning |
+| **Step 5** | Evaluate the fine-tuned model |
+| **Step 6** | Compare results |
+
+## Hardware Requirements
+
+- **Full-parameter SFT**: Qwen2.5-3B fits comfortably on a single 80 GB GPU (8 GPUs shown below for speed) with FSDP2.
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/llm/agent-sft.mdx]: We didn't do fsdp2 when it's single gpu

Comment (khazic) [reply fern/versions/nightly/pages/guides/llm/agent-sft.mdx]: Good catch — fixed in 133b3439. FSDP2 only applies across multiple GPUs, so I reworded the Hardware bullet: the single-GPU case is now described as unsharded (no FSDP2), and FSDP2 is tied to the 8-GPU example.

File: fern/versions/nightly/pages/guides/llm/agent-sft.mdx [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    dataset_name="llamafactory/glaive_toolcall_en",
+    split="train[:100]",
+    seq_length=4096,
+    truncation=True,   # see the seq_length note in Step 3
+)
+
+sample = dataset[0]
+print("keys:", list(sample.keys()))                       # input_ids, labels, attention_mask
+supervised = sum(1 for x in sample["labels"] if x != -100)
+print(f"supervised tokens: {supervised} / {len(sample['labels'])}")
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/llm/agent-sft.mdx:124]: Running the snippet (with `print(tok.decode(sample["input_ids"]))`) prints, verbatim:

```text
keys: ['input_ids', 'labels', 'attention_mask', '___PAD_TOKEN_IDS___']
supervised tokens: 127 / 502
<|im_start|>system
You are a helpful assistant.

# Tools

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
{"name": "generate_invoice", "description": "Generate an invoice for a customer", "parameters": {"type": "object", "properties": {"customer_name": {"type": "string", "description": "Name of the customer"}, "items": {"type": "array", "items": {"type": "object", "properties": {"name": {"type": "string", "description": "Name of the item"}, "quantity": {"type": "integer", "description": "Quantity of the item"}, "price": {"type": "number", "description": "Price of the item"}}, "required": ["name", "quantity", "price"]}, "description": "List of items in the invoice"}}, "required": ["customer_name", "items"]}}
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call><|im_end|>
<|im_start|>user
I need to generate an invoice for a customer named John Doe. He bought 2 apples for $1 each and 3 oranges for $2 each.<|im_end|>
<|im_start|>assistant
<tool_call>
{"name": "generate_invoice", "arguments": "{\"customer_name\": \"John Doe\", \"items\": [{\"name\": \"apple\", \"quantity\": 2, \"price\": 1}, {\"name\": \"orange\", \"quantity\": 3, \"price\": 2}]}"}
</tool_call><|im_end|>
<|im_start|>user
<tool_response>
{"invoice_id": "INV12345", "customer_name": "John Doe", "items": [{"name": "apple", "quantity": 2, "price": 1, "total": 2}, {"name": "orange", "quantity": 3, "price": 2, "total": 6}], "total_amount": 8}
</tool_response><|im_end|>
<|im_start|>assistant
The invoice for John Doe has been generated successfully. The invoice ID is INV12345. He bought 2 apples for a total of $2 and 3 oranges for a total of $6. The total amount of the invoice is $8.<|im_end|>
```

Note: the inline comment on the `keys` line (`# input_ids, labels, attention_mask`) is missing the 4th key `___PAD_TOKEN_IDS___` the dataset injects — worth fixing.

_(Posted as a plain comment, not a committable suggestion: a `suggestion` block can't contain a fenced code block, so the literal output is shown here for you to drop into the doc as an output block.)_

File: fern/versions/nightly/pages/guides/llm/agent-sft.mdx [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+}]
+messages = [{"role": "user", "content": "What's Apple's stock price right now?"}]
+
+inputs = tok.apply_chat_template(
+    messages, tools=tools, add_generation_prompt=True, return_tensors="pt", return_dict=True
+).to(model.device)
+
+with torch.inference_mode():
+    out = model.generate(**inputs, max_new_tokens=256, do_sample=False)
+print(tok.decode(out[0][inputs["input_ids"].shape[1]:], skip_special_tokens=True))
```
Comment (HuiyingLi) [inline fern/versions/nightly/pages/guides/llm/agent-sft.mdx:167]: For Step 2, the literal output of greedy decoding on base Qwen2.5-3B (`do_sample=False`, transformers 5.5.0) — exactly what gets printed:

```text
I'm looking up Apple's stock price. 🚀
<!DOCTYPE html>
<html>
<head>
  <title>Stock Price Lookup</title>
</head>
<body>
  <h1>Stock Price Lookup</h1>
  <p>Apple's stock price is $135.00.</p>
</body>
</html>

<!DOCTYPE html>
<html>
<head>
  <title>Stock Price Lookup</title>
</head>
<body>
  <h1>Stock Price Lookup</h1>
  <p>Apple's stock price is $135.00.</p>
</body>
</html>

<!DOCTYPE html>
<html>
<head>
  <title>Stock Price Lookup</title>
</head>
<body>
  <h1>Stock Price Lookup</h1>
  <p>Apple's stock price is $135.00.</p>
</body>
</html>

<!DOCTYPE html>
<html>
<head>
  <title>Stock Price Lookup</title>
</head>
<body>
  <h1>Stock Price Lookup</h1>
  <p>Apple's stock price is $135.00.</p>
</body>
</html>

<!DOCTYPE html
```

This is a good concrete illustration of the sentence right below: instead of a clean `get_stock_price(symbol="AAPL")` tool call, the base model answers in prose and then hallucinates an HTML page, repeating it until the 256-token budget runs out. Consider pasting this (or a trimmed version) under the snippet as an output block so readers see the before-fine-tuning failure concretely.

_(Plain comment, not a committable suggestion — a `suggestion` block can't contain a fenced code block. Greedy output is deterministic for a given model snapshot + transformers version, so a reader may see slight variation.)_

File: fern/versions/nightly/pages/guides/llm/agent-sft.mdx (hunk 2 above)
Comment (khazic) [reply fern/versions/nightly/pages/guides/llm/agent-sft.mdx:124]: Done in 8f67e246 — fixed the `keys` comment to include the 4th `___PAD_TOKEN_IDS___` key the dataset injects, and added a captured output block (`keys: [...]` + `supervised tokens: 127 / 502`) right under the snippet. Thanks for running it and posting the verbatim output.

File: fern/versions/nightly/pages/guides/llm/agent-sft.mdx (hunk 3 above)
Comment (khazic) [reply fern/versions/nightly/pages/guides/llm/agent-sft.mdx:167]: Done in 8f67e246 — added the greedy-decode output under the Step 2 snippet (prose + the repeated hallucinated HTML page, trimmed with a 'repeats until the token budget is exhausted' marker), plus a note that greedy output is model-snapshot / Transformers-version dependent. Thanks for capturing the literal output.

**Review by HuiyingLi — APPROVED** [review]
Thank you for the great tutorial
